### ISSUE TITLE
RUF075 false positive w/ code that should only run without exception

### ISSUE THREAD

**Issue by A5rocks** [astral-sh/ruff#25564]
### Summary

If I have some code after `yield` that I want to only run on success, RUF075 errors. Here's a concrete example of where this is useful:
```py
@asynccontextmanager
async def whatever():
  async with asyncio.TaskGroup() as tg:
    tg.start_soon(some_background_thing())
    yield  # now the background thing is running!
    tg.cancel()  # 3.15+, for demonstration
```

technically in this particular case this could also be a:
```
try:
  yield
finally:
  tg.cancel()
```

but a) that equivalence isn't always true and b) it's unnecessary (if there's an exception the task group will automatically cancel).

**Comment by ntBre** [comment]
This seems like the kind of thing it would be hard for the rule to know. Is this a property of all `asyncio.TaskGroup`s or this usage in particular? I think a `noqa` comment would be fine here if you don't want to use `try`-`finally`.

**Comment by A5rocks** [comment]
It's a property of everything with a design like TaskGroup (granted, only trio's Nurseries and asyncio's TaskGroup off the top of my head), where it will wait until completion so needs the extra cancel + automatically does so on error.

(Hopefully I'm understanding your comment correctly)

I'd be ok if the end result of this is just that I have to add noqas though.

**Comment by ntBre** [comment]
Yeah I think that was my question! I was just curious if there were a few common types that we could potentially allowlist internally, or if this only applied to using those types in a specific way. It sounds like excluding `yield`s in `asyncio.TaskGroup` or `trio.Nursery` context managers could work? I wouldn't be opposed to that.

Although I guess it would still be possible to have other code or cleanup that you'd want to run that the rule then wouldn't catch? Like if you had:

```py
def cleanup(): ...

@asynccontextmanager
async def whatever():
  async with asyncio.TaskGroup() as tg:
    tg.start_soon(some_background_thing())
    yield  # now the background thing is running!
    cleanup()
```

and an exception was raised, `cleanup` still wouldn't run right?

**Comment by A5rocks** [comment]
Yes to both questions; I think just allowing yield here would be fine + only (maybe) cancelling is valid. (One of the two cases I saw this had an `if` statement guarding it)

Not sure if that's something easily enough expressible.

**Comment by ntBre** [comment]
After seeing #27552, I'm kind of reconsidering my stance here. To bias toward false negatives over false positives, I think it could make sense to assume the surrounding context manager handles any errors, even if the `yield` isn't the last statement. But I'm curious what others think.

**Comment by tjkuson** [comment]
We can add database transactions to the list of RUF075 false-positives.

```python
import contextlib
from collections.abc import Iterator

from sqlalchemy.engine import Connection, Engine


@contextlib.contextmanager
def ctx(engine: Engine) -> Iterator[Connection]:
    with engine.connect() as connection:
        yield connection
        connection.commit()  # we don't want to commit if there's an exception!
```

If a user ends up rewriting the above using try-finally, they risk committing data they shouldn't commit.
